xclim/sdba/_adjustment.py

```diff
@@ -615,3 +615,289 @@ def extremes_adjust(
+    for i, b in enumerate(idx_binX):
+        # Get the plan row of this source bin
+        pi = np.where((b == idx_gridX).all(1))[0][0]
+        # Pick one index of this plan row
+        choice = rng.choice(range(muY.size), p=plan[pi, :])
+        out[i] = gridY[choice]
+
+    return out
+
+
+@map_groups(scen=[Grouper.DIM])
+def otc_adjust(
+    ds: xr.Dataset,
+    dim: str,
+    bin_width: list | None = None,
+    bin_origin: list | None = None,
+    numItermax: int | None = 100_000_000,
+):
+    """Optimal Transport Correction of the bias of `hist` with respect to `ref`.
+
+    Notes
+    -----
+    Dataset variables:
+      ref : training target
+      hist : training data
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        The dataset containing the data.
+    dim : str
+        The dimension along which to compute the quantiles.
+    bin_width : list | None
+        Bin widths for all dimensions.
+    bin_origin : list | None
+        Bin origins for all dimensions.
+    numItermax : int | None
+        Number of iterations of the earth mover distance algorithm.
+
+    Returns
+    -------
+    xr.Dataset
+        Adjusted data
+    """
+    dim = dim[0]
+    hist = ds.hist.dropna(dim=dim).rename(time=f"{dim}_hist")
+    ref = ds.ref.dropna(dim=dim).rename(time=f"{dim}_ref")
+    rename_kwargs = {f"{dim}_hist": dim}
+
+    scen = xr.apply_ufunc(
+        _otc_adjust,
+        hist,
+        ref,
+        kwargs=dict(
+            bin_width=bin_width,
+            bin_origin=bin_origin,
+            numItermax=numItermax,
+        ),
+        input_core_dims=[[f"{dim}_hist", "multivar"], [f"{dim}_ref", "multivar"]],
+        output_core_dims=[[f"{dim}_hist", "multivar"]],
+        keep_attrs=True,
+        vectorize=True,
+    ).rename("scen", **rename_kwargs)
+
+    return scen.to_dataset()
+
+
+def _dotc_adjust(
+    X1: np.ndarray,
+    Y0: np.ndarray,
+    X0: np.ndarray,
+    bin_width: list | None = None,
+    bin_origin: list | None = None,
+    numItermax: int | None = 100_000_000,
+    cov_factor: str | None = "std",
+):
+    """Dynamical Optimal Transport Correction of the bias of X with respect to Y.
+
+    Parameters
+    ----------
+    X1 : np.ndarray
+        Simulation data to adjust.
+    Y0 : np.ndarray
+        Bias correction reference.
+    X0 : np.ndarray
+        Historical simulation data.
+    bin_width : list | None
+        Bin widths for all dimensions.
+    bin_origin : list | None
+        Bin origins for all dimensions.
+    numItermax : int | None
+        Number of iterations of the earth mover distance algorithm.
+    cov_factor : str | None = "std"
+        Rescaling factor.
+
+    Returns
+    -------
+    np.ndarray
+        Adjusted data
+
+    References
+    ----------
+    :cite:cts:`sdba-robin_2021`
+    """
+    # Remove NaNs
+    Y0 = Y0[~np.isnan(Y0).any(axis=1), :]
+    X0 = X0[~np.isnan(X0).any(axis=1), :]
+    X1 = X1[~np.isnan(X1).any(axis=1), :]
```

Comment:
Est-ce que c'est un problème que, après cette étape, Y0, X0, et X1 peuvent avoir des grandeurs différentes? ça l'air que tu les concatines après, donc probablement. Une façon de régler c'est de prendre l'intersection des indices non-NaN pour les données, mais ça diminuera la quantité de données. Quelquechose comme ça, peut-être?

```suggestion
    nonnan_ind = ~np.isnan(Y0).any(axis=1) &&  ~np.isnan(X0).any(axis=1) &  ~np.isnan(X1).any(axis=1)
    Y0 = Y0[nonnan_ind, :]
    X0 = X0[nonnan_ind, :]
    X1 = X1[nonnan_ind, :]
```

Reply:
Je ne crois pas que ce soit un problème, les histogrammes sont standardisés de façon à sommer à 1. Je ne crois pas concaténer les données non plus, peux-tu spécifier où dans le code?

Reply:
Ah, je voyais `[Y0, X0, X1]` plus bas, mais c'est pas une concatination, juste un array de np.array